Render small-caps font variant

samizdatco opened the pull request:
- pango has a PANGO_VARIANT_SMALL_CAPS variant that can be added to the font description, but setting it doesn't cause text to be rendered in small-caps, even if the font supports it
- adding the appropriate opentype features to the layout's attributes list will use the font's small-caps glyphs if it provides them

Thanks for contributing!

- [x] Have you updated CHANGELOG.md?

samizdatco committed:
render small-caps font variant

- pango has a PANGO_VARIANT_SMALL_CAPS variant that can be added to the font description, but setting it doesn't cause text to be rendered in small-caps, even if the font supports it
- adding the appropriate opentype features to the layout's attributes list will use the font's small-caps glyphs if it provides them

samizdatco committed:
added small-caps feature to CHANGELOG

chearon commented on src/CanvasRenderingContext2d.cc:
```diff
@@ -2547,6 +2549,17 @@ NAN_SETTER(Context2d::SetFont) {
+  PangoAttribute *features;
+  if (strlen(*variant) > 0 && strcmp("small-caps", *variant)==0) {
+    features = pango_attr_font_features_new("smcp 1, onum 1");
+  }else{
```
nit: could you put spaces before/after the brackets, and spaces before/after the `==` above

samizdatco replied:
gladly

chearon reviewed (commented):
Thanks for this. Looks good in general. `pango_attr_font_features_new` is not available in the (old) version of Pango for Windows that we link to in the wiki and use in CI, so we might have to `#ifdef PANGO_VERSION(...` the new code.

We should also probably change CI testing to use MSYS like the prebuilds do.

samizdatco committed:
added textAttributes field to state struct

- the font-variant state will get out of sync with the fontDescription across context-saves & restores unless we keep a copy of the PangoAttrList around

samizdatco commented:
I realized that my initial version didn't account for state and would get out of sync with the displayed font after a save() and a restore(). So I've added a field to the state struct called `textAttributes` that holds a pointer to the layout's current `PangoAttrList` (though maybe `fontAttributes` or `layoutAttributes` would be a better name?).

samizdatco committed:
only set opentype features on sufficiently new versions of pango

- the pango NEWS file suggests 1.37.1 was where the `pango_attr_font_features_new` call was added...

samizdatco committed:
skip the small-caps variant test on windows

- it can be enabled if/when the pango version is updated

zbjornson commented on test/canvas.test.js:
```diff
@@ -447,29 +447,34 @@ describe('Canvas', function () {
 
   it('Context2d#font=small-caps', function () {
+    if (process.platform == 'win32'){
+      // the windows version of pango (< 1.37.1) is too old to support font varients
```
Doesn't have to be this PR (I can do it later unless you want to do it here) but it'd be better to expose the pango version (https://github.com/Automattic/node-canvas/blob/master/src/init.cc and https://github.com/Automattic/node-canvas/blob/master/index.js#L74) and test if it's older than 1.37.1. (Prebuilds use a newer version built from source.)

samizdatco replied:
Ah, good to know. I was wondering how to detect that directly...

samizdatco committed:
expose pango version in module

samizdatco committed:
update tests to detect pango version

chearon reviewed (commented):
Looking good, I have one comment about state management.

jeffhaack commented:
It looks like no progress on this in awhile.  Is it a complicated fix to add small-caps support in?

chearon commented:
Since #2602 we no longer use Pango. Shouldn't be terribly difficult to do in the new `FontLayout.cc`, but this no longer applies.
